File: src/hotwords.ts
import { writeFileSync } from 'fs';
import { logger } from './logger';
import { getHotwordsDB, addHotwordDB, removeHotwordDB, updateHotwordDB } from './db';

const HOTWORDS_FILE = 'hotwords.txt';

// Common English words that should never be auto-added as hotwords.
const COMMON_WORDS = new Set([
  'i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', 'your',
  'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', 'her',
  'hers', 'herself', 'it', 'its', 'itself', 'they', 'them', 'their', 'theirs',
  'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', 'these', 'those',
  'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had',
  'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if',
  'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with',
  'about', 'against', 'between', 'through', 'during', 'before', 'after', 'above',
  'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under',
  'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why',
  'how', 'all', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such',
  'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', 'too', 'very', 'can',
  'will', 'just', 'don', 'should', 'now', 'also', 'back', 'could', 'would',
  'into', 'well', 'like', 'right', 'yeah', 'okay', 'oh', 'um', 'uh', 'yes',
  'no', 'got', 'get', 'go', 'going', 'gone', 'come', 'came', 'know', 'think',
  'want', 'look', 'make', 'say', 'said', 'tell', 'told', 'take', 'took', 'see',
  'saw', 'give', 'gave', 'put', 'let', 'keep', 'kept', 'still', 'try', 'tried',
  'thing', 'things', 'something', 'nothing', 'everything', 'anything', 'someone',
  'everyone', 'anyone', 'people', 'man', 'woman', 'one', 'two', 'three', 'four',
  'five', 'first', 'last', 'new', 'old', 'good', 'bad', 'great', 'little', 'big',
  'long', 'way', 'day', 'time', 'part', 'much', 'many', 'really', 'actually',
]);

/**
 * Sync hotwords from DB to the text file that whisper_server.py reads.
 * Called after every add/remove/update so the Python server picks up changes.
 */
async function syncToFile(): Promise<void> {
  try {
    const words = await getHotwordsDB();
    writeFileSync(HOTWORDS_FILE, words.length > 0 ? words.join('\n') + '\n' : '');
  } catch (e) {
    logger.error(`Error syncing hotwords to file: ${e}`);
  }
}

export async function getHotwords(): Promise<string[]> {
  return getHotwordsDB();
}

export async function addHotword(word: string): Promise<boolean> {
  const trimmed = word.trim();
  if (!trimmed) return false;
  const added = await addHotwordDB(trimmed);
  if (added) {
    await syncToFile();
    logger.info(`Added hotword: ${trimmed}`);
  }
  return added;
}

export async function removeHotword(word: string): Promise<boolean> {
  const trimmed = word.trim();
  const removed = await removeHotwordDB(trimmed);
  if (removed) {
    await syncToFile();
    logger.info(`Removed hotword: ${trimmed}`);
  }
  return removed;
}

export async function updateHotword(oldWord: string, newWord: string): Promise<boolean> {
  const updated = await updateHotwordDB(oldWord.trim(), newWord.trim());
  if (updated) {
    await syncToFile();
    logger.info(`Updated hotword: ${oldWord} -> ${newWord}`);
  }
  return updated;
}

/** Import existing hotwords.txt into DB on first run. */
export async function migrateHotwordsToDB(): Promise<void> {
  const { existsSync, readFileSync } = await import('fs');
  const dbWords = await getHotwordsDB();
  if (dbWords.length > 0) return; // Already have data in DB

  if (existsSync(HOTWORDS_FILE)) {
    const fileWords = readFileSync(HOTWORDS_FILE, 'utf-8').split('\n').map(l => l.trim()).filter(Boolean);
    for (const word of fileWords) {
      await addHotwordDB(word);
    }
    if (fileWords.length > 0) {
      logger.info(`Migrated ${fileWords.length} hotwords from file to DB`);
    }
  }
}

/**
 * Compare original and edited transcription text. Find words in the edited
 * version that look like proper noun corrections.
 */
export function detectNewHotwords(originalText: string, editedText: string): string[] {
  const strip = (s: string) => s.replace(/[.,!?;:"'()[\]{}]/g, '');
  const origWords = new Set(originalText.split(/\s+/).map(w => strip(w).toLowerCase()));

  const candidates: string[] = [];
  for (const raw of editedText.split(/\s+/)) {
    const word = strip(raw);
    if (!word || word.length < 2) continue;
    if (word[0] !== word[0].toUpperCase() || word[0] === word[0].toLowerCase()) continue;
    if (origWords.has(word.toLowerCase())) continue;
    if (COMMON_WORDS.has(word.toLowerCase())) continue;
    candidates.push(word);
  }

  const seen = new Set<string>();
  return candidates.filter(w => {
    const lower = w.toLowerCase();
    if (seen.has(lower)) return false;
    seen.add(lower);
    return true;
  });
}

File: internal/transcribe/hotwords.go
package transcribe

import (
	"context"
	"fmt"
	"os"
	"strings"
	"unicode"

	"github.com/joe/dndscribe-go/internal/db"
)

// commonWords is a set of English words that should never be auto-added as
// hotwords from transcription corrections. Ported from dndscribe-ts.
var commonWords = map[string]bool{
	"i": true, "me": true, "my": true, "myself": true, "we": true, "our": true,
	"ours": true, "ourselves": true, "you": true, "your": true, "yours": true,
	"yourself": true, "yourselves": true, "he": true, "him": true, "his": true,
	"himself": true, "she": true, "her": true, "hers": true, "herself": true,
	"it": true, "its": true, "itself": true, "they": true, "them": true,
	"their": true, "theirs": true, "themselves": true, "what": true, "which": true,
	"who": true, "whom": true, "this": true, "that": true, "these": true,
	"those": true, "am": true, "is": true, "are": true, "was": true, "were": true,
	"be": true, "been": true, "being": true, "have": true, "has": true, "had": true,
	"having": true, "do": true, "does": true, "did": true, "doing": true,
	"a": true, "an": true, "the": true, "and": true, "but": true, "if": true,
	"or": true, "because": true, "as": true, "until": true, "while": true,
	"of": true, "at": true, "by": true, "for": true, "with": true, "about": true,
	"against": true, "between": true, "through": true, "during": true,
	"before": true, "after": true, "above": true, "below": true, "to": true,
	"from": true, "up": true, "down": true, "in": true, "out": true, "on": true,
	"off": true, "over": true, "under": true, "again": true, "further": true,
	"then": true, "once": true, "here": true, "there": true, "when": true,
	"where": true, "why": true, "how": true, "all": true, "both": true,
	"each": true, "few": true, "more": true, "most": true, "other": true,
	"some": true, "such": true, "no": true, "nor": true, "not": true,
	"only": true, "own": true, "same": true, "so": true, "than": true,
	"too": true, "very": true, "can": true, "will": true, "just": true,
	"don": true, "should": true, "now": true, "also": true, "back": true,
	"could": true, "would": true, "into": true, "well": true, "like": true,
	"right": true, "yeah": true, "okay": true, "oh": true, "um": true, "uh": true,
	"yes": true, "got": true, "get": true, "go": true, "going": true, "gone": true,
	"come": true, "came": true, "know": true, "think": true, "want": true,
	"look": true, "make": true, "say": true, "said": true, "tell": true,
	"told": true, "take": true, "took": true, "see": true, "saw": true,
	"give": true, "gave": true, "put": true, "let": true, "keep": true,
	"kept": true, "still": true, "try": true, "tried": true, "thing": true,
	"things": true, "something": true, "nothing": true, "everything": true,
	"anything": true, "someone": true, "everyone": true, "anyone": true,
	"people": true, "man": true, "woman": true, "one": true, "two": true,
	"three": true, "four": true, "five": true, "first": true, "last": true,
	"new": true, "old": true, "good": true, "bad": true, "great": true,
	"little": true, "big": true, "long": true, "way": true, "day": true,
	"time": true, "part": true, "much": true, "many": true, "really": true,
	"actually": true,
}

// SyncHotwordsFile writes the union of all campaigns' hotwords to a text file.
// WHY: The whisper engine uses this for vocabulary biasing. We write all
// campaigns because the engine is global -- it doesn't know about campaigns.
func SyncHotwordsFile(filePath string) error {
	ctx := context.Background()
	words, err := db.GetAllHotwords(ctx)
	if err != nil {
		return fmt.Errorf("get all hotwords: %w", err)
	}

	content := ""
	if len(words) > 0 {
		content = strings.Join(words, "\n") + "\n"
	}

	if err := os.WriteFile(filePath, []byte(content), 0o644); err != nil {
		return fmt.Errorf("write hotwords file: %w", err)
	}
	return nil
}

// DetectNewHotwords compares original and edited text to find new proper nouns.
// Returns candidate hotwords: capitalized words in editedText that weren't in
// originalText and aren't common English words.
func DetectNewHotwords(originalText, editedText string) []string {
	strip := func(s string) string {
		var b strings.Builder
		for _, r := range s {
			if unicode.IsPunct(r) {
				continue
			}
			b.WriteRune(r)
		}
		return b.String()
	}

	// Build set of lowercase original words.
	origWords := make(map[string]bool)
	for _, raw := range strings.Fields(originalText) {
		origWords[strings.ToLower(strip(raw))] = true
	}

	var candidates []string
	seen := make(map[string]bool)

	for _, raw := range strings.Fields(editedText) {
		word := strip(raw)
		if len(word) < 2 {
			continue
		}
		// Must start with uppercase letter.
		first := rune(word[0])
		if !unicode.IsUpper(first) {
			continue
		}
		lower := strings.ToLower(word)
		// Skip if it was already in the original text.
		if origWords[lower] {
			continue
		}
		// Skip common English words.
		if commonWords[lower] {
			continue
		}
		// Deduplicate.
		if seen[lower] {
			continue
		}
		seen[lower] = true
		candidates = append(candidates, word)
	}

	return candidates
}

// BuildInitialPrompt creates the whisper initial prompt from hotwords and
// previous context. The prompt biases whisper toward recognizing these words.
func BuildInitialPrompt(hotwords []string, previousContext string) string {
	var parts []string

	if len(hotwords) > 0 {
		parts = append(parts, "Dungeons and Dragons RPG session. Names and terms: "+strings.Join(hotwords, ", ")+".")
	} else {
		parts = append(parts, "Dungeons and Dragons RPG session with fantasy names and places.")
	}

	if previousContext != "" {
		parts = append(parts, previousContext)
	}

	return strings.Join(parts, " ")
}
